src/cmap/connection.ts

```diff
@@ -249,52 +300,335 @@ export class Connection extends EventEmitter {
-  /** @internal */
-  command(ns: string, cmd: Document, options: CommandOptions, callback: Callback): void;
   command(
-    ns: string,
+    ns: MongoDBNamespace,
     cmd: Document,
-    options: CommandOptions | Callback,
-    callback?: Callback
+    options: CommandOptions | undefined,
+    callback: Callback
   ): void {
-    wp.command(makeServerTrampoline(this), ns, cmd, options as CommandOptions, callback);
+    if (typeof ns.db === 'undefined' || typeof ns === 'string') {
+      throw new TypeError('ns cannot be a string');
+    }
+
+    const readPreference = getReadPreference(cmd, options);
+    const shouldUseOpMsg = supportsOpMsg(this);
+    const session = options?.session;
+
+    let clusterTime = this.clusterTime;
+    let finalCmd = Object.assign({}, cmd);
+    if (hasSessionSupport(this) && session) {
+      if (
+        session.clusterTime &&
+        clusterTime &&
+        session.clusterTime.clusterTime.greaterThan(clusterTime.clusterTime)
+      ) {
+        clusterTime = session.clusterTime;
+      }
+
+      const err = applySession(session, finalCmd, options as CommandOptions);
+      if (err) {
+        return callback(err);
+      }
+    }
+
+    // if we have a known cluster time, gossip it
+    if (clusterTime) {
+      finalCmd.$clusterTime = clusterTime;
+    }
+
+    if (isSharded(this) && !shouldUseOpMsg && readPreference && readPreference.mode !== 'primary') {
+      finalCmd = {
+        $query: finalCmd,
+        $readPreference: readPreference.toJSON()
+      };
+    }
+
+    const commandOptions: Document = Object.assign(
+      {
+        command: true,
+        numberToSkip: 0,
+        numberToReturn: -1,
+        checkKeys: false
+      },
+      options
+    );
+
+    // This value is not overridable
+    commandOptions.slaveOk = readPreference.slaveOk();
+    const cmdNs = `${ns.db}.$cmd`;
+    const message = shouldUseOpMsg
+      ? new Msg(cmdNs, finalCmd, commandOptions)
+      : new Query(cmdNs, finalCmd, commandOptions);
+
+    const inTransaction = session && (session.inTransaction() || isTransactionCommand(finalCmd));
+    const commandResponseHandler = inTransaction
+      ? (err?: AnyError, ...args: Document[]) => {
+          // We need to add a TransientTransactionError errorLabel, as stated in the transaction spec.
+          if (
+            err &&
+            err instanceof MongoNetworkError &&
+            !err.hasErrorLabel('TransientTransactionError')
+          ) {
+            err.addErrorLabel('TransientTransactionError');
+          }
+
+          if (
+            session &&
+            !cmd.commitTransaction &&
+            err &&
+            err instanceof MongoError &&
+            err.hasErrorLabel('TransientTransactionError')
+          ) {
+            session.transaction.unpinServer();
+          }
+
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          return callback!(err, ...args);
```

Comment:
Might not need this anymore the callback is nonnullish in this function looks like

Reply:
fixed